### PR TITLE
[PVR] rework includes

### PR DESCRIPTION
**Pull request by ksooo**
No functional changes, just cleanup of PVR includes.

@Rechi mind taking a look.

### CODE REVIEW

**Comment by ksooo** [comment]
jenkins build this please

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 1]
```diff
@@ -7,18 +7,21 @@
  */
 
 #include "Edl.h"
+
+#include "FileItem.h"
+#include "PlatformDefs.h"
+#include "ServiceBroker.h"
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/cores/VideoPlayer/Edl.cpp]: Oversight.

File: xbmc/pvr/PVRGUIDirectory.cpp [hunk 2]
```diff
@@ -8,23 +8,30 @@
...
 #include "pvr/timers/PVRTimers.h"
 #include "pvr/timers/PVRTimersPath.h"
+#include "pvr/PVRManager.h"
+#include "settings/Settings.h"
+#include "settings/SettingsComponent.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+#include "FileItem.h"
+#include "ServiceBroker.h"
+#include "utils/log.h"
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/PVRGUIDirectory.cpp]: Yes.

File: xbmc/addons/PVRClient.cpp [hunk 3]
```diff
@@ -8,41 +8,44 @@
...
+#include "utils/StringUtils.h"
+#include "utils/log.h"
+#include "ServiceBroker.h"
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+
+extern "C" {
```
Comment (Rechi) [inline xbmc/addons/PVRClient.cpp]: move curly bracket into a separate line

File: xbmc/pvr/PVRJobs.h [hunk 4]
```diff
@@ -8,17 +8,19 @@
...
 
-#include "pvr/PVRTypes.h"
+#include <memory>
+#include <string>
+#include <vector>
 
 namespace PVR
 {
+  class CPVRChannelGroup;
+  class CPVRClient;
```
Comment (Rechi) [inline xbmc/pvr/PVRJobs.h]: do not intend namespaces, https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

File: xbmc/pvr/channels/PVRChannel.h [hunk 5]
```diff
@@ -8,25 +8,23 @@
...
-class CVariant;
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
 
 namespace PVR
 {
   class CPVREpg;
+  class CPVREpgInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/channels/PVRChannel.h]: do not intend namespaces

File: xbmc/pvr/channels/PVRChannelGroupInternal.h [hunk 6]
```diff
@@ -9,11 +9,18 @@
...
+
+#include <memory>
+#include <vector>
 
 namespace PVR
 {
   enum class PVREvent;
 
+  class CPVRChannel;
+  class CPVRChannelNumber;
```
Comment (Rechi) [inline xbmc/pvr/channels/PVRChannelGroupInternal.h]: do not intend namespaces

File: xbmc/pvr/channels/PVRChannelGroups.h [hunk 7]
```diff
@@ -8,15 +8,19 @@
...
 #include "threads/SingleLock.h"
 
-#include "pvr/channels/PVRChannelGroup.h"
+#include <memory>
+#include <string>
+#include <vector>
 
 namespace PVR
 {
+  class CPVRChannel;
```
Comment (Rechi) [inline xbmc/pvr/channels/PVRChannelGroups.h]: do not intend namespaces

File: xbmc/pvr/channels/PVRChannelGroupsContainer.h [hunk 8]
```diff
@@ -8,12 +8,15 @@
...
+#include "pvr/PVRTypes.h"
 #include "threads/CriticalSection.h"
 
-#include "pvr/channels/PVRChannelGroups.h"
-
 namespace PVR
 {
+  class CPVRChannel;
+  class CPVRChannelGroups;
+  class CPVREpgInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/channels/PVRChannelGroupsContainer.h]: do not intend namespaces

File: xbmc/pvr/dialogs/GUIDialogPVRChannelManager.h [hunk 9]
```diff
@@ -8,17 +8,22 @@
...
+#include <memory>
+#include <vector>
+
+class CAction;
+class CFileItemList;
+class CGUIMessage;
 
 namespace PVR
 {
+  class CPVRClient;
```
Comment (Rechi) [inline xbmc/pvr/dialogs/GUIDialogPVRChannelManager.h]: do not intend namespaces

File: xbmc/pvr/epg/Epg.h [hunk 10]
```diff
@@ -8,20 +8,22 @@
...
+#include <memory>
+#include <string>
+#include <vector>
 
-/** EPG container for CPVREpgInfoTag instances */
 namespace PVR
 {
   class CPVREpgChannelData;
+  class CPVREpgDatabase;
+  class CPVREpgInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/epg/Epg.h]: do not intend namespaces

File: xbmc/pvr/epg/EpgContainer.h [hunk 11]
```diff
@@ -8,26 +8,30 @@
...
 
 namespace PVR
 {
-  class CPVREpgChannelData;
   class CEpgUpdateRequest;
   class CEpgTagStateChange;
+  class CPVREpg;
+  class CPVREpgChannelData;
+  class CPVREpgDatabase;
+  class CPVREpgInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/epg/EpgContainer.h]: do not intend namespaces

File: xbmc/pvr/epg/EpgInfoTag.cpp [hunk 12]
```diff
@@ -8,29 +8,38 @@
...
-: m_channelData(new CPVREpgChannelData)
+: m_iUniqueBroadcastID(EPG_TAG_INVALID_UID),
+  m_iFlags(EPG_TAG_FLAG_UNDEFINED),
+  m_channelData(new CPVREpgChannelData)
 {
 }
 
 CPVREpgInfoTag::CPVREpgInfoTag(const std::shared_ptr<CPVREpgChannelData>& channelData, int iEpgID)
-: m_iEpgID(iEpgID)
+: m_iUniqueBroadcastID(EPG_TAG_INVALID_UID),
```
Comment (Rechi) [inline xbmc/pvr/epg/EpgInfoTag.cpp]: move `,` at the beginning of the next line

File: xbmc/pvr/epg/EpgInfoTag.h [hunk 13]
```diff
@@ -8,23 +8,22 @@
...
+#include <vector>
 
-class CVariant;
+struct EPG_TAG;
+struct PVR_EDL_ENTRY;
 
 namespace PVR
 {
   class CPVREpgChannelData;
+  class CPVREpgDatabase;
```
Comment (Rechi) [inline xbmc/pvr/epg/EpgInfoTag.h]: do not intend namespaces

File: xbmc/pvr/epg/EpgInfoTag.h [hunk 14]
```diff
@@ -437,7 +436,7 @@ namespace PVR
     int                      m_iSeriesNumber = 0;   /*!< series number */
     int                      m_iEpisodeNumber = 0;  /*!< episode number */
     int                      m_iEpisodePart = 0;    /*!< episode part number */
-    unsigned int             m_iUniqueBroadcastID = EPG_TAG_INVALID_UID;   /*!< unique broadcast ID */
+    unsigned int m_iUniqueBroadcastID = 0;   /*!< unique broadcast ID */
```
Comment (Rechi) [inline xbmc/pvr/epg/EpgInfoTag.h]: remove the vertical alignment of the comment

File: xbmc/pvr/epg/EpgSearchFilter.h [hunk 15]
```diff
@@ -8,19 +8,20 @@
...
-#include "XBDateTime.h"
-
-#include "pvr/PVRTypes.h"
-#include "pvr/channels/PVRChannelNumber.h"
-
 namespace PVR
 {
   #define EPG_SEARCH_UNSET (-1)
 
+  class CPVREpgInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/epg/EpgSearchFilter.h]: do not intend namespaces

File: xbmc/pvr/recordings/PVRRecording.h [hunk 16]
```diff
@@ -24,21 +24,25 @@
...
 
 class CVideoDatabase;
-class CVariant;
+
+struct PVR_EDL_ENTRY;
+struct PVR_RECORDING;
 
 namespace PVR
 {
+  class CPVRTimerInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/recordings/PVRRecording.h]: do not intend namespaces

File: xbmc/pvr/recordings/PVRRecordings.h [hunk 17]
```diff
@@ -8,21 +8,26 @@
...
-#include "video/VideoDatabase.h"
-
-#include "pvr/PVRTypes.h"
-#include "pvr/recordings/PVRRecording.h"
+class CVideoDatabase;
 
 namespace PVR
 {
+  class CPVRRecording;
+  class CPVRRecordingUid;
```
Comment (Rechi) [inline xbmc/pvr/recordings/PVRRecordings.h]: do not intend namespaces

File: xbmc/pvr/timers/PVRTimerInfoTag.h [hunk 18]
```diff
@@ -8,33 +8,23 @@
...
+#include <memory>
+#include <string>
 
-class CVariant;
+struct PVR_TIMER;
 
 namespace PVR
 {
+  class CPVRChannel;
+  class CPVREpgInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/timers/PVRTimerInfoTag.h]: do not intend namespaces

File: xbmc/pvr/timers/PVRTimers.h [hunk 19]
```diff
@@ -8,22 +8,22 @@
...
-
-#include "pvr/PVRSettings.h"
-#include "pvr/PVRTypes.h"
-#include "pvr/timers/PVRTimerInfoTag.h"
-
 namespace PVR
 {
+  enum class TimerOperationResult;
+
+  class CPVRTimerInfoTag;
```
Comment (Rechi) [inline xbmc/pvr/timers/PVRTimers.h]: do not intend namespaces

**Comment by ksooo** [comment]
This PR is not for making PVR source files completely obeying coding guidelines. For this round I only want to fix the  include order.

File: xbmc/pvr/epg/EpgInfoTag.cpp (hunk 12 above)
Comment (ksooo) [reply xbmc/pvr/epg/EpgInfoTag.cpp]: Never ever.

**Comment by Rechi** [comment]
https://paste.kodi.tv/ipomapemub.kodi are the include changes clang-format would apply on top of your PR.

File: xbmc/pvr/channels/PVRChannelGroup.cpp [hunk 20]
```diff
@@ -11,20 +11,28 @@
 #include "PVRChannelGroup.h"
 
 #include <algorithm>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
```
Comment (Rechi) [inline xbmc/pvr/channels/PVRChannelGroup.cpp]: move C++ header includes after Kodi headers

**Comment by ksooo** [comment]
```
--- a/xbmc/addons/PVRClient.cpp
+++ b/xbmc/addons/PVRClient.cpp
@@ -8,6 +8,7 @@
 
 #include "PVRClient.h"
 
+#include "ServiceBroker.h"
 #include "addons/PVRClientMenuHooks.h"
 #include "cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h"
 #include "dialogs/GUIDialogKaiToast.h"
@@ -15,6 +16,9 @@
 #include "events/NotificationEvent.h"
 #include "filesystem/SpecialProtocol.h"
 #include "guilib/LocalizeStrings.h"
+#include "pvr/PVRDatabase.h"
+#include "pvr/PVRManager.h"
+#include "pvr/PVRStreamProperties.h"
 #include "pvr/addons/PVRClients.h"
 #include "pvr/channels/PVRChannel.h"
 #include "pvr/channels/PVRChannelGroup.h"
@@ -29,21 +33,18 @@
 #include "pvr/timers/PVRTimerInfoTag.h"
 #include "pvr/timers/PVRTimerType.h"
 #include "pvr/timers/PVRTimers.h"
-#include "pvr/PVRDatabase.h"
-#include "pvr/PVRManager.h"
-#include "pvr/PVRStreamProperties.h"
 #include "settings/AdvancedSettings.h"
 #include "settings/SettingsComponent.h"
 #include "utils/StringUtils.h"
 #include "utils/log.h"
-#include "ServiceBroker.h"
 
 #include <map>
 #include <memory>
 #include <string>
 #include <utility>
 
-extern "C" {
+extern "C"
+{
 #include <libavcodec/avcodec.h>
 }
 ```
@Rechi wtf? It seems clang-format wants files before directories, which I prefer and had used before, but coding guidelines states "Place directories before files.". That's what I did. So, waht is correct?

**Comment by Rechi** [comment]
> Place directories before files.

I didn't know that this was in the coding guidelines.
clang-format doesn't move files before directories but does a case sensitive sorting (uppercase before lower case).

**Comment by ksooo** [comment]
>> Place directories before files.

> I didn't know that this was in the coding guidelines.

 And it makes no sense, imo. @pkerling what do you think, should we change coding guidelines to "case sensitive sorting (uppercase before lower case)"?

File: xbmc/addons/PVRClient.cpp (hunk 3 above)
Comment (ksooo) [reply xbmc/addons/PVRClient.cpp]: done

File: xbmc/pvr/channels/PVRChannelGroup.cpp (hunk 20 above)
Comment (ksooo) [reply xbmc/pvr/channels/PVRChannelGroup.cpp]: done

**Comment by ksooo** [comment]
I reshuffled includes again. Now  clang-format should be happy.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
jenkins build this please

File: xbmc/pvr/PVRGUIDirectory.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (fuzzard) [reply xbmc/pvr/PVRGUIDirectory.cpp]: Wrong order on utils/log.h

File: xbmc/addons/PVRClient.cpp [hunk 21]
```diff
@@ -24,25 +16,37 @@ extern "C" {
...
 #include "pvr/epg/EpgInfoTag.h"
+#include "pvr/recordings/PVRRecording.h"
 #include "pvr/recordings/PVRRecordings.h"
 #include "pvr/timers/PVRTimerInfoTag.h"
 #include "pvr/timers/PVRTimerType.h"
 #include "pvr/timers/PVRTimers.h"
+#include "settings/AdvancedSettings.h"
+#include "settings/SettingsComponent.h"
+#include "utils/StringUtils.h"
+#include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/addons/PVRClient.cpp]: log.h is out of order in a few places when other util headers are included.. I assume it's doing some.ordering that includes case sensitivity

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 22]
```diff
@@ -7,18 +7,21 @@
...
 #include "settings/SettingsComponent.h"
-#include "utils/log.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
 #include "utils/XBMCTinyXML.h"
-#include "PlatformDefs.h"
-#include "pvr/recordings/PVRRecordings.h"
-#include "pvr/PVRManager.h"
-#include "ServiceBroker.h"
+#include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/cores/VideoPlayer/Edl.cpp]: log.h order again

File: xbmc/pvr/PVRGUIActions.cpp [hunk 23]
```diff
@@ -21,43 +21,57 @@
...
 #include "pvr/windows/GUIWindowPVRSearch.h"
+#include "settings/MediaSettings.h"
+#include "settings/Settings.h"
+#include "threads/IRunnable.h"
+#include "threads/SingleLock.h"
+#include "utils/StringUtils.h"
+#include "utils/SystemInfo.h"
+#include "utils/URIUtils.h"
+#include "utils/Variant.h"
+#include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/pvr/PVRGUIActions.cpp]: log.h order

File: xbmc/pvr/PVRGUIChannelIconUpdater.cpp [hunk 24]
```diff
@@ -19,12 +17,15 @@
...
 #include "settings/Settings.h"
 #include "settings/SettingsComponent.h"
-#include "utils/URIUtils.h"
-#include "utils/log.h"
-
 #include "pvr/PVRGUIProgressHandler.h"
 #include "pvr/channels/PVRChannel.h"
 #include "pvr/channels/PVRChannelGroup.h"
+#include "utils/URIUtils.h"
+#include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/pvr/PVRGUIChannelIconUpdater.cpp]: log.h order

File: xbmc/pvr/PVRManager.cpp [hunk 25]
```diff
@@ -8,35 +8,38 @@
...
 #include "pvr/recordings/PVRRecordings.h"
-#include "pvr/recordings/PVRRecordingsPath.h"
+#include "pvr/timers/PVRTimerInfoTag.h"
 #include "pvr/timers/PVRTimers.h"
+#include "settings/Settings.h"
+#include "utils/JobManager.h"
+#include "utils/Stopwatch.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+#include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/pvr/PVRManager.cpp]: log.h order

File: xbmc/pvr/PVRSettings.cpp [hunk 26]
```diff
@@ -10,16 +10,19 @@
...
 #include "guilib/LocalizeStrings.h"
+#include "pvr/PVRGUIActions.h"
+#include "pvr/PVRManager.h"
+#include "pvr/addons/PVRClients.h"
 #include "settings/Settings.h"
 #include "settings/SettingsComponent.h"
-#include "settings/lib/SettingDefinitions.h"
 #include "settings/lib/SettingsManager.h"
 #include "utils/StringUtils.h"
 #include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/pvr/PVRSettings.cpp]: Assume best to change this order as well

File: xbmc/pvr/channels/PVRChannel.cpp [hunk 27]
```diff
@@ -9,17 +9,22 @@
...
+#include "pvr/PVRDatabase.h"
+#include "pvr/PVRManager.h"
+#include "pvr/epg/Epg.h"
+#include "pvr/epg/EpgChannelData.h"
+#include "pvr/epg/EpgContainer.h"
+#include "pvr/epg/EpgInfoTag.h"
 #include "threads/SingleLock.h"
 #include "utils/StringUtils.h"
 #include "utils/Variant.h"
 #include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/pvr/channels/PVRChannel.cpp]: Again, maybe change

File: xbmc/pvr/dialogs/GUIDialogPVRClientPriorities.cpp [hunk 28]
```diff
@@ -7,17 +7,16 @@
...
-
-#include <cstdlib>
 
 #include "ServiceBroker.h"
+#include "guilib/GUIMessage.h"
+#include "pvr/PVRManager.h"
 #include "settings/lib/Setting.h"
 #include "utils/StringUtils.h"
-#include "utils/Variant.h"
 #include "utils/log.h"
```
Comment (fuzzard) [inline xbmc/pvr/dialogs/GUIDialogPVRClientPriorities.cpp]: Here again, I'll stop going now, I'm sure you get the point.

**Comment by fuzzard** [comment]
Just saw the comments about clang-tidy doing case sensitive ordering. Maybe better to change utils/log.h to utils/Log.h, but obviously that will touch on a lot more than just pvr, so feel free to disregard my comments

File: xbmc/addons/PVRClient.cpp (hunk 21 above)
Comment (ksooo) [reply xbmc/addons/PVRClient.cpp]: Nope, that's fine. upper case letters < lower case letters.

**Comment by ksooo** [comment]
> so feel free to disregard my comments

 done. ;-)

File: xbmc/pvr/PVRGUIChannelIconUpdater.cpp [hunk 29]
```diff
@@ -19,12 +17,15 @@
 #include "settings/AdvancedSettings.h"
 #include "settings/Settings.h"
 #include "settings/SettingsComponent.h"
-#include "utils/URIUtils.h"
-#include "utils/log.h"
-
 #include "pvr/PVRGUIProgressHandler.h"
 #include "pvr/channels/PVRChannel.h"
 #include "pvr/channels/PVRChannelGroup.h"
```
Comment (Rechi) [inline xbmc/pvr/PVRGUIChannelIconUpdater.cpp]: pvr before settings

File: xbmc/pvr/channels/PVRChannelGroup.cpp [hunk 30]
```diff
@@ -10,21 +10,28 @@
 
 #include "PVRChannelGroup.h"
 
-#include <algorithm>
-
 #include "ServiceBroker.h"
+#include "addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h"
+#include "addons/PVRClient.h"
```
Comment (Rechi) [inline xbmc/pvr/channels/PVRChannelGroup.cpp]: PVRClient.h before kodi-addon-dev-kit

File: xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp [hunk 31]
```diff
@@ -8,18 +8,21 @@
 
 #include "GUIDialogPVRGuideSearch.h"
 
-#include <utility>
-
 #include "ServiceBroker.h"
-#include "addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h"
 #include "guilib/GUIMessage.h"
 #include "guilib/GUIEditControl.h"
```
Comment (Rechi) [inline xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp]: GUIEditControl.h before GUIMessage.h

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp [hunk 32]
```diff
@@ -17,18 +18,24 @@
...
+#include "input/actions/ActionIDs.h"
 #include "pvr/PVRGUIActions.h"
 #include "pvr/PVRGUIDirectory.h"
 #include "pvr/PVRManager.h"
-#include "pvr/addons/PVRClients.h"
 #include "pvr/channels/PVRChannelGroup.h"
+#include "pvr/channels/PVRChannelGroups.h"
 #include "pvr/channels/PVRChannelGroupsContainer.h"
+#include "messaging/ApplicationMessenger.h"
+#include "messaging/helpers/DialogOKHelper.h"
```
Comment (Rechi) [inline xbmc/pvr/windows/GUIWindowPVRBase.cpp]: messaging before pvr

File: xbmc/pvr/windows/GUIWindowPVRTimers.cpp [hunk 33]
```diff
@@ -10,13 +10,12 @@
 
 #include "FileItem.h"
 #include "utils/URIUtils.h"
-
 #include "pvr/timers/PVRTimersPath.h"
```
Comment (Rechi) [inline xbmc/pvr/windows/GUIWindowPVRTimers.cpp]: pvr before utils

File: xbmc/pvr/PVRGUIChannelIconUpdater.cpp (hunk 29 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIChannelIconUpdater.cpp]: done

File: xbmc/pvr/channels/PVRChannelGroup.cpp (hunk 30 above)
Comment (ksooo) [reply xbmc/pvr/channels/PVRChannelGroup.cpp]: done

File: xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp (hunk 31 above)
Comment (ksooo) [reply xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp]: done

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp (hunk 32 above)
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: done

File: xbmc/pvr/windows/GUIWindowPVRTimers.cpp (hunk 33 above)
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRTimers.cpp]: done

**Comment by ksooo** [comment]
jenkins build this please

**Pull request merged**